src/Annotation/AbstractAnnotation.php

```
@@ -0,0 +1,90 @@
+<?php
+
+namespace Opstalent\CrudBundle\Annotation;
+
+/**
+ * Class AbstractAnnotation
+ * @author Szyon Kunowski <szymon.kunowski@gmail.com>
+ * @package Opstalent\CrudBundle
+ */
+abstract class AbstractAnnotation
+{
+    /**
+     * @var string[]
+     *
+     * @Required
+     * @AnnotatedDescription("The property holds all available actions for class")
+     */
+    public $actions = [];
+
+    /**
+     * @var array
+     */
+    protected $availableActions = ['addable', 'getable', 'listable', 'editable'];
+
+    /**
+     * @param string $key
+     * @param array $data
+     * @return array
+     */
+    protected function setActions(string $key, array $data): AbstractAnnotation
+    {
+        if (array_key_exists($key, $data)) {
+            foreach ($data[$key] as $action) {
+                $this->enableAction($action);
+            }
+        }
+        return $this;
+    }
+
+    /**
+     * @return array
+     */
+    public function getActions(): array
+    {
+        return $this->actions;
+    }
+
+    /**
+     * @return array
+     */
+    protected function getAvailableActions(): array
+    {
+        return $this->availableActions;
+    }
+
+    /**
+     * @param string $name
+     * @return int
+     */
+    protected function addAvailableAction(string $name): AbstractAnnotation
+    {
+        array_push($this->availableActions, $name);
+        return $this;
+    }
+
+    /**
+     * @param string $action
+     * @return bool
+     */
+    protected function isActionAvailable(string $action): bool
+    {
+        if (in_array($action, $this->availableActions)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @param string $action
+     * @return Entity
+     */
+    protected function enableAction(string $action): AbstractAnnotation
+    {
+        if ($this->isActionAvailable($action)) {
+            array_push($this->actions, $action);
+        }
+        return $this;
```

what to do if it's not available? ignoring?

---

why not ? should we throw exception ? we could but what we achieve with that ?

---

yes, enabling not allowed action should throw an exception